elrond-adapter/aggregator/exchange.go

```diff
@@ -0,0 +1,126 @@
+package aggregator
+
+import (
+	"log"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/ElrondNetwork/elrond-adapter/config"
+)
+
+const (
+	QuoteFiat   = "USD"
+	QuoteStable = "USDT"
+)
+
+const (
+	minValidResults = 3
+)
+
+type Exchange interface {
+	FetchPrice(base, quote string) (float64, error)
+}
+
+type PairData struct {
+	Base            string
+	Quote           string
+	ScAddress       string
+	Endpoint        string
+	PriceMultiplied string
+}
+
+type ExchangeAggregator struct {
+	exchanges []Exchange
+	prices    map[string]float64
+	config    config.ExchangeConfig
+}
+
+var supportedExchanges = []Exchange{
+	&Binance{},
+	&Bitfinex{},
+	&Cryptocom{},
+	&Kraken{},
+	&Gemini{},
+	&Huobi{},
+	&Hitbtc{},
+	&Okex{},
+}
+
+func NewExchangeAggregator(exchangeConfig config.ExchangeConfig) *ExchangeAggregator {
+	prices := make(map[string]float64)
+	for _, pair := range exchangeConfig.Pairs {
+		prices[pair.Base] = 0
+	}
+	return &ExchangeAggregator{
+		exchanges: supportedExchanges,
+		config:    exchangeConfig,
+		prices:    prices,
+	}
+}
+
+func (eh *ExchangeAggregator) GetPricesForPairs() []PairData {
+	var results []PairData
+	for _, pair := range eh.config.Pairs {
+		currPrice, err := eh.GetPrice(pair.Base, pair.Quote)
+		if err != nil {
+			log.Println(err)
+			break
+		}
+
+		lastPrice := eh.prices[pair.Base]
+		pairData := PairData{
+			Base:            pair.Base,
+			Quote:           pair.Quote,
+			ScAddress:       pair.ScAddress,
+			Endpoint:        pair.Endpoint,
+			PriceMultiplied: eh.MultiplyFloat64CastStr(currPrice),
+		}
+
+		if lastPrice == 0 || !eh.config.CheckPercentageChange {
+			results = append(results, pairData)
+		} else {
+			percentageChange := PercentageChange(currPrice, lastPrice)
+			if percentageChange >= eh.config.PercentageThreshold {
+				results = append(results, pairData)
+			}
```

Comment:
`lastPrice` (which is set above) should be set only if the `PercentageThreshold` check passes... otherwise a stock which changes gradually will not be updated at all even if the total cumulated changes may exceed the `PercentageThreshold`

Reply:
`lastPrice` is first read from the map at line `71` and is updated after each iteration at line `88`. Updating only when the threshold check is met would result in price mismatch, because it would load a historical price, not the last fetched. Also if `lastPrice` is not updated every time means the else branch is never reached.

Reply:
Updating the `lastPrice` every time means you discard the percentage changes.
Let's say the `config.PercentageThreshold` is `2.5%` but you have a relatively stable currency that changes with `1%` several times in a row (50 times for example's sake).
By setting `lastPrice` to the most recent value even though an update wasn't sent means that you accumulate `50 * 1% = 50%` change which is greater than the `2.5%` threshold, but will not trigger an append to `results`.

Reply:
got it. good catch. pushed a fix